presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -975,12 +974,9 @@ else if (criteria instanceof JoinOn) {
 
                 Analyzer.verifyNoAggregatesOrWindowFunctions(metadata.getFunctionRegistry(), expression, "JOIN clause");
 
-                Expression canonicalized = CanonicalizeExpressions.canonicalizeExpression(expression);
-                analyzeExpression(canonicalized, output);
-
                 Set<Expression> postJoinConjuncts = new HashSet<>();
 
-                for (Expression conjunct : ExpressionUtils.extractConjuncts(canonicalized)) {
+                for (Expression conjunct : ExpressionUtils.extractConjuncts(expression)) {
                     conjunct = ExpressionUtils.normalize(conjunct);
```

Comment:
`analysis.recordSubqueries(node, analyzeExpression(postJoinPredicate, output));` does not register coercions in `analysis`. It only registers subqueries like the name says.